### PR TITLE
Coding conventions and contributing guidelines

### PR DESCRIPTION
**Pull request by AnakinRaW**

### CODE REVIEW

File: CODING.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Coding Guidelines
+
+## General Notes
+In general we aim for keeping the coding guidelines that are common by the Java Language. This includes code style,
```
Comment (krinara86) [inline CODING.md:4]: common "for"

File: CODING.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+as well as the usage of coding patterns. Exceptions have to get discusses and agreed with the project leads.
+
+## Checkstyle Config
+This repository contains a `checkstyl.xml` Checkstyle config which gets automatically used by the CI Runner for code
```
Comment (krinara86) [inline CODING.md:8]: checkstyle.xml

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  - You can skip this step for trivial changes.
+  - Reuse an existing issue on the topic, if there is one.
+  - If you want to implement the issue yourself, please state this. 
+2. Create a personal fork of the repository on GitHub. 
+3. In your fork, create a new branch of develop (`git checkout -b mybranch`).
+    - Apply branch naming rules (see above)
+3. Make and commit your changes to your branch.
+    - Keep changes as small as possible.
+    - Try to follow the [Commit Messages](#commit-messages) guidance.
+4. Add new tests corresponding to your change, if applicable.
```
Comment (krinara86) [inline CONTRIBUTING.md:33]: 4.1: If you donot add tests, explain why they are not required in the PR.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+## Branches
+
+This repository contains two central branches. The **master** and the **develop** branch. The develop branch is default.
+Both branches are *protected* against direct write access thus Pull Requests are necessary to push into them. 
+Other branches are unprotected and can be created and deleted by an contributer
+
+The `master` branch holds the lastest stable release of the application.
+
+The `develop` branch holds the lastest development version of the application
```
Comment (krinara86) [inline CONTRIBUTING.md:55]: I would add one more line on the expected workflow. For example, how and when is development synchronized into master.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+feature/    // For new features
+fix/        // Fixes a bug
+hotfix/     // Fixes a critical bug in a release build
+migration/  // Moves code from the experiment to application code
+revert/     // Reverts a commit
+```
+
+## Commit Messages
+
+Try to format commit messages as follows (based on [A Note About Git Commit Messages](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)):
```
Comment (krinara86) [inline CONTRIBUTING.md:80]: In the commit message, use the first word to describe what component you are changing. For example, 

"utils: creating new ceiling function"

File: CODING.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# Coding Guidelines
+
+## General Notes
+In general we aim for keeping the coding guidelines that are common by the Java Language. This includes code style,
+as well as the usage of coding patterns. Exceptions have to get discusses and agreed with the project leads.
```
Comment (schlichtig) [inline CODING.md:5]:
```suggestion
as well as the usage of coding patterns. Exceptions have to get discussed and agreed with the project leads.
```

File: CODING.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+## Checkstyle Config
+This repository contains a `checkstyl.xml` Checkstyle config which gets automatically used by the CI Runner for code 
+validation on the **master**, **develop** and **Pull Requests** branches.
+
+Feature branches do not checked because we want to things getting done quickly. This can be done if everybody codes as
```
Comment (schlichtig) [inline CODING.md:11]:
```suggestion
Feature branches are not automatically checked because we want to get development done quickly. To achieve this, everybody may codes as 
```

File: CODING.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+as well as the usage of coding patterns. Exceptions have to get discusses and agreed with the project leads.
+
+## Checkstyle Config
+This repository contains a `checkstyl.xml` Checkstyle config which gets automatically used by the CI Runner for code 
+validation on the **master**, **develop** and **Pull Requests** branches.
+
+Feature branches do not checked because we want to things getting done quickly. This can be done if everybody codes as 
+they feel most familiar. A pull request will make sure `develop` and `master` branches are consistent.   
+
+The checkstyle config got confirmed by the core development team at UPB and IEM. Changes may only be made by the project leaders
```
Comment (schlichtig) [inline CODING.md:14]: UPB, IEM and TUD  ??

File: CODING.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+with serious reasons.
```
Comment (schlichtig) [inline CODING.md:15]:
```suggestion
for valid reasons.
```

File: CODING.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+### Tool Support
+
+Checkstyle is available for various IDEs.
+
+##### Eclipse
+1. Install checkstyle plugin from [here](https://checkstyle.org/eclipse-cs/)
+2. In your preferences window you can select the checkstyle configuration `checkstyle.xml` from the root of the
+repository.
+3. Configure the Eclipse Code Formatter to automatically format the code regarding the checkstyle rules. 
+Unfortunately it does not support importing the Checkstyle Rules directly.
```
Comment (schlichtig) [inline CODING.md:26]: Capitalization of Checkstyle (Rules) is inconsistent

File: CODING.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+##### IntelliJ
+1. Search for "Checkstyle-IDEA" in the `Settings->Plugin` window.
+2. Install and restart the IDE.
+3. Open settings again and select the `iem_checks.xml` file in `Settings-Checkstyle`.
```
Comment (schlichtig) [inline CODING.md:31]: Do we maybe want to rename the xml-files into something CogniCrypt-related (for clarity)?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+- **DO** follow the naming conventions for branches.
+- **DO** include tests when adding new features. When fixing bugs, start with adding a test that highlights how the current behavior is broken.
+- **DO** keep the discussions focused. When a new or related topic comes up it's often better to create new issue than to side track the discussion.
+- **DO** document your changes in detail in the pull requests.
+
+Please do not:
+- **DON'T** make PRs for coding guidelines changes.
+- **DON'T** create a big pull requests with no related issue. Instead, first file an issue.
+- **DON'T** submit PRs that use licensed files we are not allowed to use.
+- **DON'T** add or change API related code without filing an issue and discussing with the project leads first
```
Comment (schlichtig) [inline CONTRIBUTING.md:17]:
```suggestion
- **DON'T** add or change API related code without filing an issue and discussing it with the project leads first
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+8. Wait for feedback or approval of your changes.
+9. When remaining issues from feedback are resolved and all checks are green, your PR will be merged by one of the reviewers.
+    - Delete the source branch with the merge.
+
+
+## Branches
+
+This repository contains two central branches. The **master** and the **develop** branch. The develop branch is default.
+Both branches are *protected* against direct write access thus Pull Requests are necessary to push into them. 
+Other branches are unprotected and can be created and deleted by an contributer
```
Comment (schlichtig) [inline CONTRIBUTING.md:51]:
```suggestion
Other branches are unprotected and can be created and deleted by a contributer
```

File: checkstyle.xml [hunk 14]
```diff
@@ -0,0 +1,425 @@
+<?xml version="1.0"?>
```
Comment (schlichtig) [inline checkstyle.xml:1]: @AnakinSklavenwalker Do you maybe want to add the changes we discussed in the meeting?

File: CONTRIBUTING.md (hunk 4 above)
Comment (AnakinSklavenwalker) [reply CONTRIBUTING.md:56]: That's something we have not yet decided

File: checkstyle.xml (hunk 14 above)
Comment (AnakinSklavenwalker) [reply checkstyle.xml:1]: iirc i already added them

File: .github/workflows/build_and_test.yml [hunk 15]
```diff
@@ -14,9 +14,6 @@ jobs:
   build-and-test-analysis:
     runs-on: ubuntu-latest
 
-    # Do not run this build in the merge queue
-    if: github.event_name != 'merge_group'
```
Comment (schlichtig) [inline .github/workflows/build_and_test.yml:18]: Is this removal intended?

File: .github/workflows/build_and_test.yml [hunk 16]
```diff
@@ -43,9 +42,6 @@ jobs:
   build-and-test-scanner-soot:
     runs-on: ubuntu-latest
 
-    # Do not run this build in the merge queue
-    if: github.event_name != 'merge_group'
```
Comment (schlichtig) [inline .github/workflows/build_and_test.yml:47]: Same here?

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+  - You can skip this step for trivial changes.
+  - Reuse an existing issue on the topic, if there is one.
+  - If you want to implement the issue yourself, please state this. 
+2. Create a personal fork of the repository on GitHub. 
+3. In your fork, create a new branch of develop (`git checkout -b mybranch`).
+    - Apply branch naming rules (see below)
+3. Make and commit your changes to your branch.
+    - Keep changes as small as possible.
+4. Add new tests corresponding to your change, if applicable.
+  - Minor changes, like fixing typos, adding documentation or non-critical bugfixes may are excluded.
```
Comment (schlichtig) [inline CONTRIBUTING.md:33]: 'may be'

**Review by schlichtig — CHANGES_REQUESTED** [review]
Please address the small comments

**Review by schlichtig — APPROVED** [review]
